### ISSUE TITLE
Syntax for Rich Text & Dynamic Templates

### ISSUE THREAD

**Issue by jeromesimeon** [accordproject/cicero#392]
This issue is related to #346 but with a specific focus on syntax for discussion. Previous related issue #171 is closed and somewhat outdated, so I think a new issue is useful.

## Requirements and proposals

The requirements for rich text and dynamic templates syntax are:
1. Handle rich-text annotations, based as much as possible on some existing spec https://github.com/accordproject/cicero/issues/392#issuecomment-511441589
2. Handle arbitrary switching between Ergo expressions and text https://github.com/accordproject/cicero/issues/392#issuecomment-511442767
3. Handle variables, expressions, lists, tables, based as much as possible on some existing template mechanism https://github.com/accordproject/cicero/issues/392#issuecomment-511497777
4. Support both parse and generateText methods in Cicero https://github.com/accordproject/cicero/issues/392#issuecomment-511502307

Some eye can be kept on backward compatibility, although this is not a strong requirement and consistency of design may take precedence.

## Due diligence

- Directly related issues https://github.com/accordproject/cicero/issues/392#issuecomment-511510057
- Breaking changes https://github.com/accordproject/cicero/issues/392#issuecomment-511514658

**Comment by jeromesimeon** [comment]
For 1. Handle rich-text annotations, the main candidate is markdown based on common mark. which is already used in the Markdown Editor used by Cicero UI.

Some open questions:
- Support for tables?
- Should XML-like markup be allowed or not?
- How does it interact with the dynamic template support?

**Comment by jeromesimeon** [comment]
For 2. Current prototype uses `{{ ... }}` for both switching from text to expression, and from expression to text. Some alternatives:

- `This is text {{ let x = 1 in {{ This is text again with value {{ x }}}} }}` (as current prototype)
- `This is text {{ let x = 1 in `This is text again with value {{ x }}` }}` (more like JavaScript templates strings)
- ...

**Comment by mttrbrts** [comment]
I'm intrigued by the `\`` notation, for completeness and consistency with the conceptual use in JS, here's a third option:
```
This is text ` let x = 1; {{ This is text again with value ` x `}} ` 
```

**Comment by jeromesimeon** [comment]
> I'm intrigued by the ` notation, for completeness and consistency with the conceptual use in JS, here's a third option:
> 
> ```
> This is text ` let x = 1; {{ This is text again with value ` x `}} ` 
> ```

The correspondence with code blocks is intriguing...

**Comment by jeromesimeon** [comment]
For 3. A proposal inspired by [handlebars](https://handlebarsjs.com). _Note: In the range between very little language-binding (mustache) to deep integration with a language (JSX), I find handlebars the most relevant/close to what we are trying to achieve_

## Variables

```
{{# firstName }} // Source variable (used for parsing and rendering)
{{ firstName }}  // Variable used for rendering
```

## Ergo Expressions

Many special purpose things in handlebars can be written as Ergo expressions:

```
{{ author.name }}                    // Nested paths (as paths)
{{ roundn(rate,precision) }}         // Helpers (as function calls)
{{ roundn(rate,2) }}                 // Literals (also as function calls)
{{ foreach p in products
   return `Product: {{ p.name }}` }} // Iterators
{{ if state = "US"
   then address.state
   else address.country }}           // Conditionals
{{ if forceMajeure
   then `This is a force majeure` }} // More conditionals
```

Other things you can do with Ergo expressions:

```
{{ firstName ++ lastName }}                // String concatenation
{{ enforce (rate > 0.0) }}                 // Check conditions
etc.
```

## Block Expressions

### Lists

```
{{#list products}}Product {{name}} with code {{id}}{{/list}}
```

Those are rendered as markdown. From:
```
{ items: [
    { name: "Sponge", id: 1312 },
    { name: "Bleach", id: 341 }
  ] }
```
to
```
- Product "Sponge" with code 1312
- Product "Bleach" with code 341
```

You can change the delimiter:
```
{{#list(",") products}}{{name}}{{/list}}
```
to render:
```
"Sponge","Bleach"
```

### Clauses

Clauses can be inlined:
```
Payment
-------
{{#clause payment}}
As consideration in full for the rights granted herein, Licensee shall pay Licensor a one-time fee in the amount of [{amountText}] ([{amount}]) upon execution of this Agreement, payable as follows: {[paymentProcedure]}.
[{/clause}]
```

Or can be imported:
```
Payment
-------
{{> ap://payment-upon-delivery@0.9.0.cta}}
```

### Tables

**TBD**

## Other things that could be included

### Comments

```
{{!-- This is a comment --}}
```

### Whitespace control

```
{{~name}}  // Remove whitespace on left
{{name~}}  // Remove whitespace on right
{{~name~}} // Remove whitespace on both
```

### Iterators Blocks

Eventhough you could use an Ergo expression, you could also support:
```
{{#foreach products}}
- Product: {{ name }}
{{/foreach}}

{{#foreach(",") products}}Product: {{ name }}{{/foreach}}
```

### Conditionals Blocks

Eventhough you could use an Ergo expression, you could also support:
```
{{#if forceMajeure}}
This is a force majeure
{{/if}}

{{#if isActive}}
  I'm active
{{else}}
  I'm not active
{{/if}}
```

**Comment by jeromesimeon** [comment]
For 4.

The idea is to identify the source variables in the text, and generate a parser when possible. What _could_ handled by the parser is:
```
[{variableName}]  // Source variables
[{variableName.fieldName}] // Nested variables
```

Those can still occur inside the following blocks:

```
{{#list}}...{{/list}}           // List blocks may context parsable variables
{{#with fieldName}}...{{/with}} // With blocks may contain parsable variables
```

**Comment by jeromesimeon** [comment]
Issues directly addressed by the proposal:
https://github.com/accordproject/cicero/issues/296
https://github.com/accordproject/cicero/issues/346
https://github.com/accordproject/cicero/issues/79
https://github.com/accordproject/cicero/issues/20
https://github.com/accordproject/cicero/issues/17
https://github.com/accordproject/cicero/issues/4

Relevant issues that are ***not*** addressed by the proposal:
https://github.com/accordproject/cicero/issues/19
https://github.com/accordproject/cicero/issues/16

**Comment by jeromesimeon** [comment]
Breaking changes from our current syntax:
- Variables should be changed from `[{variableName}]` to `{[variableName]}`
- Formatted variables should be changed from `[{variableName as "FORMAT"}]` to `{[variableName as "FORMAT"]}`
- Boolean variables should be changed from `[{"This is a force majeure":?forceMajeure}]` to `{{#if forceMajeure}}This is a force majeure{{/if}}`
- Nested clauses should be changed from `[{#paymentClause}]As consideration in full for the rights granted herein...[{/paymentClause}]` to `{{#clause payment}}As consideration in full for the rights granted herein...{{/clause}}`

**Comment by jeromesimeon** [comment]
Alternative Syntax For _3. Handle variables, expressions, lists, tables, based as much as possible on some existing template mechanism_

The proposal bellow is inspired by [handlebars](https://handlebarsjs.com). 

> Note: In the range between very little language-binding (mustache) and deep integration with a language (JSX), I find handlebars the most relevant/close to what we are trying to achieve

## Variables ✅

Variables are quoted with `{{...}}`.

```
{{firstName}}                       // Source variable (used for parsing and rendering)
{{deliveryDate as "MMMM, DD YYYY"}} // Source variable with date formatting
```

> `[` may be an issue for the common mark parser because it is used for links.

## Ergo Expressions ✅

Expressions are quoted with `{{% ... %}}`

Many special purpose things in handlebars can be written as Ergo expressions:

```
{{% author.name %}}                    // Nested paths (as paths)
{{% roundn(rate,precision) %}}         // Helpers (as function calls)
{{% roundn(rate,2) %}}                 // Literals (also as function calls)
{{% foreach p in products
    return `Product: {{ p.name }}` %}} // Iterators
{{% if state = "US"
    then address.state
    else address.country %}}           // Conditionals
{{% if forceMajeure
    then `This is a force majeure` %}} // More conditionals ❌
```

Other things you can do with Ergo expressions:

```
{{% firstName ++ lastName %}}                // String concatenation
{{% enforce rate > 0.0 %}}                   // Enforce conditions ❌
etc.
```

## Block Expressions

Block expressions of kind `kind` with scope `scope` and parameters `params` are quoted with `{{#kind scope params}} ... {{/kind}}`.

### Conditionals Blocks ✅

```
{{#if forceMajeure}}
This is a force majeure
{{/if}}

{{#if isActive}}
  I'm active
{{else}}
  I'm not active
{{/if}}
```

### Unordered Lists ✅

```
{{#ulist products}}Product {[name}} with code {[id}}{{/ulist}}
```

Those are rendered as markdown. From:
```
{ products: [
    { name: "Sponge", id: 1312 },
    { name: "Bleach", id: 341 }
  ] }
```
to
```
- Product "Sponge" with code 1312
- Product "Bleach" with code 341
```

### Ordered Lists ✅

```
{{#olist products}}Product {{name}} with code {{id}}{{/olist}}
```

Those are rendered as markdown. From:
```
{ products: [
    { name: "Sponge", id: 1312 },
    { name: "Bleach", id: 341 }
  ] }
```
to
```
1. Product "Sponge" with code 1312
1. Product "Bleach" with code 341
```

### Joined Lists ✅

You create a list with a delimiter
```
{{#join products ","}}{{name}}{{/join}}
```

To render from:
```
{ products: [
    { name: "Sponge", id: 1312 },
    { name: "Bleach", id: 341 }
  ] }
```
to render:
```
"Sponge","Bleach"
```

### Clauses

Clauses can be inlined: ✅
```
Payment
-------
{{#clause payment}}
As consideration in full for the rights granted herein, Licensee shall pay Licensor a one-time fee in the amount of {{amountText}} ({{amount}}) upon execution of this Agreement, payable as follows: {{paymentProcedure}}.
{{/clause}}
```

Or can be imported: ❌
```
Payment
-------
{{>payment ap://payment-upon-delivery@0.9.0.cta}}
```

### Tables ❌

_Note: This is a very tentative part of the proposal_

```
{{#table products}}
  {{#cell "Name"}}{{name}}{{/cell}}
  {{#cell "Code"}}{{id}}{{/cell}}
{{/table}}
```

Those are rendered as markdown. From:
```
{ products: [
    { name: "Sponge", id: 1312 },
    { name: "Bleach", id: 341 }
  ] }
```
to
```
| Name | Code |
|---|---|
| "Sponge" | 1312 |
| "Bleach" | 341 |
```

## Other things that could be included

### Iterators Blocks ❌

Eventhough you could use an Ergo expression, you could also support:
```
{{#foreach products}}
- Product: {{name}}
{{/foreach}}

{{#foreach(",") products}}Product: {{name}}{{/foreach}}
```

### With ✅

With which can be used to set a context:
```
{{#with author}}
{{firstName}} {{lastName}}
{{/with}}
```

instead of

```
{{% author.firstName %}} {{% author.lastName %}}
```

### Comments ❌

```
{{!-- This is a comment --}}
```

### Whitespace control ❌

```
{{~name}}  // Remove whitespace on left
{{name~}}  // Remove whitespace on right
{{~name~}} // Remove whitespace on both
```

**Comment by jeromesimeon** [comment]
Alternative Syntax For _2. Handle arbitrary switching between Ergo expressions and texts_

Expressions inside of texts are quoted with `{{% ... %}}`. Text inside expressions are anti-quoted using `%{{ ... }}%`.

For instance:

```
Exhibit B
Authorized Licensed Products and Territory
Products:
{{% foreach p in products return %{{- {{% p.productDescription %}}\n}}% %}}
Territory:
{{% foreach c in countries return %{{- {{% c %}}\n}}% %}}
```

**Comment by jeromesimeon** [comment]
This proposal has been implemented as part of release `0.20`. Specific features that did not make the `0.20` but could be of interest have been opened as new issues. I'm closing this.
